File: 23. Merge k Sorted Lists/mergeKLists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class mergeKLists {
    public ListNode merge2Lists(ListNode l1, ListNode l2) {
        ListNode head = new ListNode(-1);
        ListNode dummy = head;
        while (l1 != null && l2 != null) {
            if (l1.val < l2.val) {
                dummy.next = l1;
                l1 = l1.next;
                dummy = dummy.next;
            }
            else {
                dummy.next = l2;
                l2 = l2.next;
                dummy = dummy.next;
            }
        }
        while (l1 != null) {
            dummy.next = l1;
            l1 = l1.next;
            dummy = dummy.next;
        }
        while (l2 != null) {
            dummy.next = l2;
            l2 = l2.next;
            dummy = dummy.next;
        }
        return head.next;
    }

    public ListNode merge(ListNode[] lists, int start, int end) {
        if (start == end) {
            return lists[start];
        }
        int mid = (start + end) / 2;
        ListNode l1 = merge(lists, start, mid);
        ListNode l2 = merge(lists, mid + 1, end);
        return merge2Lists(l1, l2);
    }

    public ListNode solution(ListNode[] lists) {
        if (lists.length == 0) {
            return null;
        }
        return merge(lists, 0, lists.length - 1);
    }
}

File: 23. Merge k Sorted Lists/mergeKLists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class mergeKLists:
    def merge2Lists(self, l1, l2):
        if not l1:
            return l2
        if not l2:
            return l1
        if l1.val < l2.val:
            l1.next = self.merge2Lists(l1.next, l2)
            return l1
        else:
            l2.next = self.merge2Lists(l1, l2.next)
            return l2

    def merge(self, lists: List[ListNode], start, end):
        if start == end:
            return lists[start]
        mid = (start + end) // 2
        l1 = self.merge(lists, start, mid)
        l2 = self.merge(lists, mid + 1, end)
        return self.merge2Lists(l1, l2)

    def solution(self, lists: List[ListNode]) -> ListNode:
        if not lists:
            return []
        return self.merge(lists, 0, len(lists) - 1)

